extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitter.java

```diff
@@ -207,6 +207,15 @@ public void close()
     } else {
       exec.shutdownNow();
       flush();
+
+      if (pushGateway != null && config.isDeleteOnShutdown()) {
+        try {
+          pushGateway.delete(config.getNamespace(), ImmutableMap.of(config.getNamespace(), identifier));
```

Comment:
Will this operation delete _all_ task metrics, including any non-stale metrics that were recently flushed above? What do you think about leveraging `push_time_seconds` added by the PushGateway and use that to cleanup only stale metrics that are older than a defined TTL. Something like:
- Query prometheus for `push_time_seconds{identifier}`  - this will require additional prometheus server config, though.
- Then delete the identifier metrics only if the timestamp is older than say 1 minute.

Reply:
@BartMiki - thanks for the explanation. Digging up some more, it seems like the pushgateway community considers this usecase an [anti-pattern](https://github.com/prometheus/pushgateway/tree/master#non-goals). Also, from their [mailing](https://groups.google.com/g/prometheus-developers/c/9IyUxRvhY7w) thread, it appears that different folks have implemented their custom TTL expiration strategies. So I'm wondering if any of the available open source tools/solutions from the thread will solve the staleness problem - instead of implementing a TTL-based strategy that is generally discouraged by the prometheus pushgateway community? @kfaraz, do you've any thoughts here?

Reply:
@abhishekrb19 Thank you for the research on this topic! 

The TTL is discouraged by implementing it as part of Pushgateway, however, the solution they present in the Google Groups thread is the solution I described above, mainly ([link](https://groups.google.com/g/prometheus-developers/c/9IyUxRvhY7w/m/edX2nq3_BQAJ)):
> Solution: Do _not_ implement a TTL. Instead, make metrics
deletion (with a simple DELETE call of the RESTful API) a part of
decommissioning a job.

My approach is basically that. After the Druid task closes, do not remove the metrics from PushGateway immediately, but schedule a task on Druid to call the removal of these metrics after some time elapses (TTL but it is on the Druid side). That would be part of decommissioning a job as per their recommendation.

Also, could you elaborate on what is considered an anti-pattern here? You write here:
> Digging up some more, it seems like the pushgateway community considers this usecase an [anti-pattern](https://github.com/prometheus/pushgateway/tree/master#non-goals).

I'm not sure about the use case you are referring to. To my understanding, the Pushgateway on Druid should be only used for Peons, which aligns with the PushGateway [goals](https://github.com/prometheus/pushgateway/tree/master#prometheus-pushgateway):
> The Prometheus Pushgateway exists to allow ephemeral and batch jobs to expose their metrics to Prometheus. Since these kinds of jobs may not exist long enough to be scraped, they can instead push their metrics to a Pushgateway. The Pushgateway then exposes these metrics to Prometheus.

Reply:
Hi @abhishekrb19 and @kfaraz!  Should I proceed with the proposed change? What is your opinion? Also regarding the build errors, the build passes on my machine perhaps the test rerun would help (I have had a similar issue with other PR too and the rerun has helped).

Reply:
Sorry for the delayed review! :( 

Thinking more, the missing metrics problem exists with the Prometheus `exporter` strategy (the pull model) - i.e., when the Druid servers are shutting down, and Prometheus hasn't scraped the metrics yet on time.

I think what is proposed here suffices, and @BartMiki and I discussed this offline. In addition to the patch, he suggested adding a delay config, something like `waitForShutdownDelay`, so the peon tasks will wait for the configured time before the delete is issued at the time of shutdown, which seems reasonable.